pkg/util/runtime/features.go

```diff
@@ -0,0 +1,98 @@
+// Copyright 2020 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package runtime
+
+import (
+	"net/url"
+	"strconv"
+
+	"github.com/pkg/errors"
+	"github.com/spf13/pflag"
+	"github.com/spf13/viper"
+)
+
+const (
+	featureGateFlag = "feature-gates"
+
+	// FeatureExample is an example feature gate flag, used for testing and demonstrative purposes
+	FeatureExample Feature = "Example"
+)
+
+var (
+	featureGates    map[Feature]bool
+	featureDefaults = map[Feature]bool{
+		FeatureExample: false,
+	}
+)
+
+// Feature is a type for defining feature gates.
+type Feature string
+
+// FeaturesBindFlags does the Viper arguments configuration. Call before running pflag.Parse()
+func FeaturesBindFlags() {
+	viper.SetDefault(featureGateFlag, "")
+	pflag.String(featureGateFlag, viper.GetString(featureGateFlag), "Flag to pass in the url qyery list of feature flags to enable or disable")
+}
+
+// FeaturesBindEnv binds the environment variables, based on the flags provided.
+// call after viper.SetEnvKeyReplacer(...) if it is being set.
+func FeaturesBindEnv() error {
+	return viper.BindEnv(featureGateFlag)
+}
+
+// ParseFeaturesFromEnv will parse the feature flags from the Viper args
+// configured by FeaturesBindFlags() and FeaturesBindEnv()
+func ParseFeaturesFromEnv() error {
+	return ParseFeatures(viper.GetString(featureGateFlag))
+}
+
+// ParseFeatures parses the url encoded query string of features and stores the value
+// for later retrieval
+func ParseFeatures(queryString string) error {
+	features := map[Feature]bool{}
+	values, err := url.ParseQuery(queryString)
+	if err != nil {
+		return errors.Wrap(err, "error parsing query string for feature gates")
+	}
+
+	for k := range values {
```

Comment:
Should this do any error checking on the value of the feature name? A simple check would be length (e.g. not more than 64 chars) or if we wanted to be more pedantic you could verify that the feature name was something hard-coded into this file (in a list of features that we know about).

Reply:
I like the idea of throwing an error if the feature gate is not a valid one in the list -- will help with user error. Adding this in now.